Use cron schedule for defragmentation

swapnilgm opened the pull request:
Signed-off-by: Swapnil Mhamane <swapnil.mhamane@sap.com>

**What this PR does / why we need it**:
This PR changes the defragmentation period configuration to cron like schedule. Also, partially addresses #125 
**Which issue(s) this PR fixes**:
Fixes partially https://github.com/gardener/etcd-backup-restore/issues/125

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator|developer
-->
```action user
Defragmentation schedule can be configured now in cron standards using flag `defragmentation-schedule`. :warning: Removed the flag `defragmentation-period-hours`.
```
```action user
:warning: Removed the flag `delta-snapshot-period-seconds`. Instead use replacement flag `delta-snapshot-period` with input value format supported by golang `time.Duration`.
```
```action user
:warning: Removed the flag `garbage-collection-period-seconds`. Instead use replacement flag `garbage-collection-period` with input value format supported by golang `time.Duration`.
```

swapnilgm commented:
@shreyas-s-rao I have rebased the PR. This PR is now ready for the review. While refactoring, i have added the changes for garbage-collection-period and delta-snapshot-period in same PR. Sorry to have it in single PR itself.  
I would also prefer to have this PR in, before rest of PR just to minimise the efforts with rebasing and conflicts.

shreyas-s-rao commented on pkg/snapshot/restorer/restorer_test.go:
```diff
@@ -343,42 +332,34 @@ var _ = Describe("Running Restorer", func() {
+				go utils.PopulateEtcdWithWaitGroup(populatorCtx, wg, logger, endpoints, nil)
+				defer cancelPopulator()
+				ssrCtx := utils.ContextWithWaitGroupFollwedByGracePeriod(testCtx, wg, time.Second)
+				time.Sleep(2) // This is to make sure that
```
```suggestion
				time.Sleep(2) // This is to make sure that populator stops before running snapshotter
```

swapnilgm replied:
Not needed. Removed sleep.

shreyas-s-rao commented on pkg/snapshot/snapshotter/snapshotter_test.go:
```diff
@@ -117,9 +118,9 @@ var _ = Describe("Snapshotter", func() {
-				endpoints = []string{"http://localhost:5000"}
-				etcdConnectionTimeout = 5
+				validEndpoint := etcd.Clients[0].Addr().String()
+				endpoints = []string{validEndpoint[:len(validEndpoint)-2]}
```
This might fail if the valid endpoint itself is below 100 (it'll have only 2 digits). Maybe it's better to convert port number to int and add 1 to it. WDYT?

swapnilgm commented on test/utils/utils.go:
```diff
@@ -0,0 +1,176 @@
+
+// ContextWithGracePeriod returns a new context, whose Done channel is closed when parent
+//context is closed with additional <gracePeriod>.
+func ContextWithGracePeriod(parent context.Context, gracePeriod time.Duration) context.Context {
```
I am following the naming convention from context package here.

swapnilgm commented on test/utils/utils.go:
```diff
@@ -0,0 +1,176 @@
+
+// ContextWithWaitGroupFollwedByGracePeriod returns a new context, whose Done channel is closed
+// when parent context is closed or wait of waitGroup is over, with additional <gracePeriod>.
+func ContextWithWaitGroupFollwedByGracePeriod(parent context.Context, wg *sync.WaitGroup, gracePeriod time.Duration) context.Context {
```
See comment above.

swapnilgm commented on test/utils/utils.go:
```diff
@@ -0,0 +1,176 @@
+					logger.Infof("Populated data till key %s into embedded etcd with etcd end revision: %v", KeyPrefix+strconv.Itoa(response.KeyTo), response.EndRevision)
+					return
+				}
+				response.EndRevision = resp.Header.GetRevision()
```
Nope. Precisely that why the `if response.KeyTo > keyTo ` is at the start of case.  Don't have to handle it at multiple place. I am updating the function comment to describe the behaviour. Also, for now divisible factor `10` kepy constant. We will make it as parameter when required. WDYS?

shreyas-s-rao replied:
Sure, makes sense. But then the log message at line 112 should be changed to print `keyTo` or `response.KeyTo-1` instead of `response.KeyTo`.

swapnilgm commented on test/utils/utils.go:
```diff
@@ -0,0 +1,176 @@
+			logger.Infof("Populated data till key %s into embedded etcd with etcd end revision: %v", KeyPrefix+strconv.Itoa(response.KeyTo), response.EndRevision)
+			return
+		case <-time.After(time.Second):
+			response.KeyTo++
```
See comment below.

swapnilgm commented on test/utils/utils.go:
```diff
@@ -0,0 +1,176 @@
+			return
+		case <-time.After(time.Second):
+			response.KeyTo++
+			if response.KeyTo > keyTo {
```
see comment below.

swapnilgm commented on pkg/etcdutil/defrag_test.go:
```diff
@@ -98,30 +104,28 @@ var _ = Describe("Defrag", func() {
+			cancelStatusReq()
 			Expect(err).ShouldNot(HaveOccurred())
 
 			Expect(defragCount).Should(BeNumerically("==", expectedDefragCount))
```
Ahh sorry make sense. Will update it.

swapnilgm commented:
Thank you for the review @shreyas-s-rao. PTAL. I have addressed the comments. PTAL.

> Edit: While running the test, I see the following message for the defrag spec should defragment and reduce size of DB within time:

{"level":"warn","ts":"2019-09-25T13:53:04.466+0530","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"passthrough:///127.0.0.1:60091","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}

This is due to vendor update. Retry_interceptor.go from etcd has log message irrespective of error. The error appears on etcd server stop call. So, nothing to worry about it.
